File: Sem3/Paradigms/lab2/lab2_scala/src/main/scala/zad2.scala
import scala.annotation.tailrec

@tailrec
def length[A](arr: List[A], acc: Int = 0): Int = {
  if (arr.isEmpty) acc
  else length(arr.tail, acc + 1)
}

@main def main(): Unit = {
  val test1 = length(List(5, 4, 3, 2)) == 4
  val test2 = length(List("a", "b", "c")) == 3
  val test3 = length(List()) == 0
  val test4 = length(List(true, false, true, false, true)) == 5
  val test5 = length((1 to 1000000).toList) == 1000000

  println(s"Test 1: $test1")
  println(s"Test 2: $test2")
  println(s"Test 3: $test3")
  println(s"Test 4: $test4")
  println(s"Test 5: $test5")
}

File: Sem3/Paradigms/lab4/lab4_ocaml/zad2.ml
let bin_to_dec bits = 
  let rec aux bits acc =
    match bits with
    | [] -> acc
    | h :: t -> aux t (acc * 2 + h)
  in
  aux bits 0

let test1 = bin_to_dec [1;0;0;1] = 9
let test2 = bin_to_dec [0;0;0;0] = 0
let test3 = bin_to_dec [1;1;1;1;1] = 31
let test4 = bin_to_dec [] = 0
let test5 = bin_to_dec [1;0;0;0;0;0;0;0] = 128
let test6 = bin_to_dec [1] = 1
let test7 = bin_to_dec [0] = 0

let _ =
  print_endline (string_of_bool test1);
  print_endline (string_of_bool test2);
  print_endline (string_of_bool test3);
  print_endline (string_of_bool test4);
  print_endline (string_of_bool test5);
  print_endline (string_of_bool test6);
  print_endline (string_of_bool test7);